HeterogeneousCore/CUDAServices/src/CUDAService.cc

```diff
@@ -67,6 +67,7 @@ CUDAService::~CUDAService() {
     // Explicitly destroys and cleans up all resources associated with the current device in the
     // current process. Any subsequent API call to this device will reinitialize the device.
     // Useful to check for memory leaks with `cuda-memcheck --tool memcheck --leak-check full`.
+    cudaCheck(cudaDeviceSynchronize());
     cudaCheck(cudaDeviceReset());
```

Comment:
This may be for a subsequent PR, but I suppose in the presence of multiple devices this would need a loop over the devices.

Reply:
Yes.

Also, it doesn't actually fix the problem :-(

Migrating all calls from the device to the runtime API does not fix t either :-(

Still investigating...